libsignal-service/src/models.rs

```diff
@@ -1,33 +1,75 @@
+use crate::{proto::Verified, ParseServiceAddressError, ServiceAddress};
+
+use bytes::Bytes;
+use thiserror::Error;
+
 /// Attachment represents an attachment received from a peer
-///
-/// Source: `textsecure/textsecure.go`
+#[derive(Debug)]
 pub struct Attachment<R> {
+    pub content_type: String,
     pub reader: R,
-    pub mime_type: String,
 }
 
-/// Group holds group metadata
-///
-/// Source: `textsecure/groups.go`
-pub struct Group {
-    pub id: Vec<u8>,
-    pub hex_id: String,
-    pub flags: u32,
+/// Mirror of the protobuf ContactDetails message
+/// but with stronger types (e.g. `ServiceAddress` instead of optional uuid and string phone numbers)
+/// and some helper functions
+#[derive(Debug)]
+pub struct Contact {
+    pub address: ServiceAddress,
     pub name: String,
-    pub members: Vec<String>,
-    pub avatar: Option<Vec<u8>>,
+    pub color: Option<String>,
+    pub verified: Verified,
+    pub profile_key: Vec<u8>,
+    pub blocked: bool,
+    pub expire_timer: u32,
+    pub inbox_position: u32,
+    pub archived: bool,
+    pub avatar: Option<Attachment<Bytes>>,
```

Comment:
I wonder whether it makes sense to really store this as bytes, instead of a pointer to where the attachment is stored?

Reply:
well, if you look at the `Attachment` struct, it can be anything that implements `Reader` - in this particular context (contacts attachment from a sync message), avatars are always binary data you get directly without the need to fetch it.